<link rel="stylesheet" type="text/css" href="css/help.css">
<link rel="stylesheet" type="text/css" href="lib/github.css">
<script src="lib/highlight.pack.js"></script>
<script>
hljs.initHighlightingOnLoad();
</script>
<script>
$(document).ready(function(){
	renderAllApiDocs();
})

function loadScript(scriptPath, obj){
	$.ajax({
		url: scriptPath,
		dataType: 'text',
		success: function(script){
			var suffix = scriptPath.split('.')[1];
			console.log(suffix)
			switch (suffix){
				case 'py':
					var scriptType = 'python';
					break;
				case 'json':
					var scriptType = 'json';
					break;
				default:
					var scriptType = 'r';
			}
			var codeBlock = $('<code>');
			codeBlock = codeBlock.html(script).addClass(scriptType);
			codeBlock = $('<pre>').append(codeBlock)
			obj.append(codeBlock);
		}
	});
}

function renderParams(data){
	var table = $('<table>').addClass('table parameters').append('<tbody>');
	var params = data["Parameters"]
	for (var i = 0; i < params.length; i++) {
		var param = params[i];
		var tr = $('<tr>');
		for (var i = 0; i < param.length; i++) {
			tr.append( $('<td>').html(param[i]) )
		};
		table.append(tr)
	};
	return table;
}

function renderScripts(data){
	var navTabs = $('<ul>').addClass('nav nav-tabs');
	var tabContents = $('<div>').addClass('tab-content');
	scriptPaths = data['Example code'];

	i = 0;
	for (var key in scriptPaths){
		var scriptPath = scriptPaths[key];
		var id = scriptPath.replace('.', '-');

		var li = $('<li>').attr('data-toggle', 'tab')
			.attr('role', 'presentation')
			.attr('data-target', '#'+id);
		li.append($('<a>').html(key));

		var tabPane = $('<div>').attr('id', id)
			.addClass('tab-pane fade');

		if (i === 0){
			li = li.addClass('active');
			tabPane = tabPane.addClass('active in');
		}

		navTabs = navTabs.append(li);
		tabContents = tabContents.append(tabPane);
		i++;
	}
	return {navTabs: navTabs, tabContents: tabContents};
}

function renderApiDoc(data){
	FIELDS = [
		'Method',
		'URL',
		'Returns',
		'Parameters',
		'Example code',
		'Example result',
		];
	var li = $('<li>');
	var h4 = $('<h4>').html(data['name']);
	var dl = $('<dl>').addClass('dl-horizontal');
	li = li.append(h4)
	
	for (var i = 0; i < FIELDS.length; i++) {
		var field = FIELDS[i];
		var dt = $('<dt>').html(field);
		var dd = $('<dd>')
		switch (field){
			case 'Parameters':
				dd.append(renderParams(data));
				break;
			case 'Example code':
				scriptDOMs = renderScripts(data)
				dd.append(scriptDOMs['navTabs']);
				dd.append(scriptDOMs['tabContents']);
				break;
			case 'Example result':
				dd.attr('id', data['Example result'].replace('.', '-'));
				break;
			default:
				dd.html(data[field]);
		}

		dl.append(dt);
		dl.append(dd);
	};

	li = li.append(dl);
	return li;
}

function renderAllApiDocs(){
	$.getJSON('api_docs.json', function(apiDocs){
		console.log(apiDocs)
		for (var i = 0; i < apiDocs.length; i++) {
			var li = renderApiDoc(apiDocs[i]);
			$('#api-docs').append(li);
			// append scripts after json got
			for (var key in apiDocs[i]['Example code']){
				var scriptPath = apiDocs[i]['Example code'][key];
				var id = scriptPath.replace('.', '-');
				console.log(scriptPath)
				loadScript(scriptPath, $('#' + id));
			}
			// append result
			var resultPath = apiDocs[i]['Example result'];
			var id = resultPath.replace('.', '-')
			loadScript(resultPath, $('#' + id))
			
		};
	});	
}

</script>

<div class="container-fluid">
	<div class="row">

		<div class="col-md-10 col-md-offset-1">
			<div class="page-header" id="/#contribution">
				<h3>Instructions for contributing gene expression signatures</h3>
			</div>
			<div>

				<h4>Find studies of interest on <a href="http://www.ncbi.nlm.nih.gov/geo/">GEO</a></h4>
				<ul>
					<li>The search term can be name of genes, drugs or diseases.</li>
					<li>The study should be restricted to mammals, including human, mouse and rat.</li>
					<li>An example search: <a target="_blank" href="http://www.ncbi.nlm.nih.gov/gds/?term=(TP53)+AND+(%22homo+sapiens%22%5BOrganism%5D+OR+%22mus+musculus%22%5BOrganism%5D+OR+%22rattus+norvegicus%22%5BOrganism%5D)">(TP53) AND ("homo sapiens"[Organism] OR "mus musculus"[Organism] OR "rattus norvegicus"[Organism])</a></li>
				</ul>
				
				<h4>Use <a href="http://amp.pharm.mssm.edu/g2e/">GEO2Enrichr</a> to extract meta-data and identify differentially expressed genes</h4>
				<ul>
					<li>GEO2Enrichr is a browser extension for analyzing gene expression data on GEO</li>
					<li>Detailed instruction for installing and using the software is available in <a href="http://amp.pharm.mssm.edu/g2e/manual">this manual</a></li>
				</ul>

				<h4>Submit signatures to the <a href="http://maayanlab.net/crowdsourcing/index.php">BD2K-LINCS-DCIC Crowdsourcing Portal</a></h4>
				<ul>
					<li>User login is required for submission. You can either register for an account or log in through Google account.</li>
					<li>Choose a relevant microtask for the gene expression signature you collected.</li>
					<li>Fill out the form and submit. </li>
					<li><a href="http://maayanlab.net/crowdsourcing/microtask_leaderboard.php">Leaderboard</a> is available to checkout number of submissions for each users.</li>
				</ul>

			</div>

			<div class="page-header" id="/#api-doc">
				<h3>API Documentation</h3>
			</div>
			<div>
				<ul id="api-docs"></ul>
			</div>
		</div>

	</div>
</div>		